src/geb/geb.lisp

```diff
@@ -157,3 +157,49 @@ Example:
         ((distribute x y z)  (coprod (prod x y) (prod x z)))
         ((inject-left x y)   (coprod x y))
         ((inject-right x y)  (coprod x y)))))
+
+(defun gather (x y z)
+  (pair (mcase (<-left x y) (<-left x z))
+        (mcase (comp (->left y z) (<-right x y))
+               (comp (->right y z) (<-right x z)))))
+
+(defun prod-left-assoc (f)
+  (trivia:match (dom f)
+    ((prod w (prod x y))
+     (comp f (pair (comp (<-left w x) (<-left (prod w x) y))
+                   (pair (comp (<-right w x) (<-left (prod w x) y))
+                         (<-right (prod w x) y)))))
+    (_ (error "object ~A need to be of a double product type, however it is of ~A" f (dom f)))))
+
+(defun right (f)
+  (trivia:match (dom f)
+    ((coprod x y)
+     (comp f (->right x y)))
+    (_ (error "object ~A need to be of a coproduct type, however it is of ~A"
+              f (dom f)))))
+
+(defun left (f)
+  (trivia:match (dom f)
+    ((coprod x y)
+     (comp f (->left x y)))
+    (_ (error "object ~A need to be of a coproduct type, however it is of ~A"
+              f (dom f)))))
+
+(defgeneric curry (f)
+  (:documentation "Curries the given object"))
+
+(defmethod curry ((f <substmorph>))
+  (cond ((not (typep f 'substmorph))
+         (subclass-responsibility f))
+        ((not (typep (dom f) 'prod))
+         (error "object ~A need to be of a product type, however it is of ~A" f (dom f)))
+        (t
+         (let ((dom (dom f)))
+           (match-of substobj (mcadr dom)
+             (alias        (curry f))
+             (so0          (terminal (mcadr dom)))
+             (so1          (comp f (pair (mcadr dom)
+                                         (terminal (mcar dom)))))
+             ((coprod x y) (pair (curry (left (comp f (gather (mcar dom) x y))))
```

Comment:
seems like
`(comp f (gather (mcar dom) x y))` 

is a pattern, maybe we can `flet` it out, not really needed here, but might reduce on the repeat logic, though if this is a deeper idea, maybe we should make it a top level function?